File: OneQuestionPerDay/LCCI17/get_kth_magic_number.go
package main

func getKthMagicNumber(k int) int {
	dp := make([]int, k)
	dp[0] = 1
	three := 0
	five := 0
	seven := 0

	for i := 1; i < k; i++ {
		dp[i] = min3(3*dp[three], 5*dp[five], 7*dp[seven])
		if 3*dp[three] == dp[i] {
			three++
		}
		if 5*dp[five] == dp[i] {
			five++
		}
		if 7*dp[seven] == dp[i] {
			seven++
		}
	}
	return dp[k-1]
}

func min3(x, y, z int) int {
	return min(x, min(y, z))
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

File: OneQuestionPerDay/LCCI17/get_kth_magic_number.py
class Solution:
    def getKthMagicNumber(self, k: int) -> int:
        dp = [1] * k
        p3 = p5 = p7 = 0

        for i in range(1, k):
            f3 = 3 * dp[p3]
            f5 = 5 * dp[p5]
            f7 = 7 * dp[p7]
            dp[i] = min(f3, f5, f7)
            if f3 == dp[i]:
                p3 += 1
            if f5 == dp[i]:
                p5 += 1
            if f7 == dp[i]:
                p7 += 1
        return dp[k - 1]
